### ISSUE TITLE
Simplify promotion after A/B(/n) test

### ISSUE THREAD

**Issue by kalantar** [iter8-tools/iter8#1375]
**Is your feature request related to a problem? Please describe.**
Promoting a winning version requires several steps to avoid dropping user requests:
- unset `iter8.tools/abn` label on default version to step traffic to it
- upgrade the default version
- set `iter8.tools/abn` label to `true` to restart traffic to (upgraded) default version
- unset `iter8.tools/abn` on the candidate version
- delete candidate resources

Iter8 is not responsible for these steps. However, it is reasonable to ask: is there something Iter8 can do that will reduce the effort for the user?

**Describe the solution you'd like**
Enable simpler user interaction for promotion, perhaps as simple as:

- upgrading the default version and 
- deleting the candidate version

**Other Information**
Most of the user effort is in checking for version readiness and setting the label `iter8.tools/abn`.
If Iter8 could compute version readiness itself, then the user interaction would be simplified to just 2 steps as desired.

To compute readiness, the Iter8 service needs to know, in advance, all of the resources that comprise a version since it is necessary to determine when all of the necessary resources are available. Iter8 must be able to identify for each application/track pair the list of resources (GVR, namespace/name)

The user can provide this list as part of the configuration of the Iter8 Service. For example, the following configuration might suffice:

```yaml
ns:
  hello:
    baseline:
    - group: apps
      version: v1
      resource: deployments
      name: hello-deployment
    - version: v1
      resource: services
      name: hello-svc
    candidate:
```

User configuration can be simplified if several simplifying assumptions are made:
- All versions of an application are deployed to the same namespace
- Tracks have names are fixed as: baseline, candidate-0, candidate-1, etc 
- There is a single resource of each type
- Resource names have the form <app_name>-<resource_type>
- There is a known mapping between a set of “resource types” to GVR

Then the user configuration might be as simple as:
--set ns1.hello=service-deploy \
--set ns2.goodbye.knative=goodbye

**Comment by kalantar** [comment]
Since user specifies full list of resources for each version, the labels expected by Iter8 can be reduced. Only the version label is still required. The app and track labels are part of the specification and the ready label is computed.

**Comment by kalantar** [comment]
Given the increased detail that is part of specification, permissions needed by the Iter8 service can be more specific. It is no longer necessary to give permission to get/list all resources of specific types. The permissions can be restricted to specific instances.

Furthermore, since the current service is not aware of the applications in advance, the service needs permission to get, list, create and update (all) secrets (used to store metric values). Now the app names are predefined allowing (a) more targeted permissions and (b) pre-creation of the metric secrets.

**Comment by kalantar** [comment]
Given a change in resource (ie, triggered by informer), we
- using name/namespace lookup configuration to identify app/track and from that all resources
- Get these resources from the cluster (if some missing, not ready)
- Evaluate for readiness
- If all ready then record as ready
- otherwise, not ready

To evaluate readiness:
```
for all resources r specified for track:
  get resource
  if not present → return not ready
  if being deleted → return not ready
  return isReady (r)
```

A resource is marked with a `deletionTimestamp` when it is in the process of being deleted allowing us to identify resources being deleted.

The `isReady()` method must implement logic specific to each resource type. Ideally, share logic with existing readiness task.

**Comment by kalantar** [comment]
Resource naming should be modified. It is suggested above that the name be <app_name>-<resource_type>.  This is insufficient to distinguish between resources deployed for different versions. Instead suggest we assume:

- All resources for all versions of an application are deployed in the same namespace
- There is only 1 resource of a given type in a version
- All resources in the default or baseline version are named using the application name
- All resources in a candidate version are named by appending “-candidate-i” to the application name where i is an index

It is then possible to configure the application by identifying, for each app, the namespace, the app name and a list of resource types (which can be mapped to GVR).  For example:

```shell
helm install --repo https://iter8-tools.github.io/hub iter8-abn iter8-abn \
   --set "default.backend.resources={service,deployment}"
#   --set default.backend.maxNumCandidates=1

```

This config should work:
```yaml
default:
  backend:
    maxNumCandidates: 1
    resources:
    - version: v1
      resource: services
    - group: apps
      version: v1
      resource: deployments
```

**Comment by kalantar** [comment]
This has been released in v0.13
